// vuepress-plugin-right-anchor/lib/node/rightAnchorPlugin.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.rightAnchorPlugin = void 0;
const utils_1 = require("@vuepress/utils");
const rightAnchorPlugin = (options = {}) => {
    return {
        name: 'vuepress-plugin-right-anchor',
        clientAppRootComponentFiles: utils_1.path.resolve(__dirname, '../client/components/RightAnchor.js'),
        extendsPageData: (page) => {
            const { rightAnchor: frontmatterOptions = {} } = page.frontmatter;
            const rightAnchor = {
                ...options,
                ...frontmatterOptions,
                isIgnore: Array.isArray(options.ignore) && options.ignore.includes(page.path),
                expand: {
                    trigger: 'hover',
                    clickModeDefaultOpen: true,
                    ...options.expand,
                    ...frontmatterOptions.expand,
                },
            };
            return {
                rightAnchor
            };
        },
    };
};
exports.rightAnchorPlugin = rightAnchorPlugin;
