// @vuepress/client/lib/setupGlobalComputed.js
import { computed } from 'vue';
import { pageData, pageFrontmatterSymbol, pageHeadSymbol, pageHeadTitleSymbol, pageLangSymbol, routeLocaleSymbol, siteData, siteLocaleDataSymbol, } from './composables';
import { withBase } from './helpers';
import { resolvers } from './resolvers';
/**
 * Create and provide global computed
 */
export const setupGlobalComputed = (app, router) => {
    // create global computed
    const routeLocale = computed(() => resolvers.resolveRouteLocale(siteData.value.locales, router.currentRoute.value.path));
    const siteLocaleData = computed(() => resolvers.resolveSiteLocaleData(siteData.value, routeLocale.value));
    const pageFrontmatter = computed(() => resolvers.resolvePageFrontmatter(pageData.value));
    const pageHeadTitle = computed(() => resolvers.resolvePageHeadTitle(pageData.value, siteLocaleData.value));
    const pageHead = computed(() => resolvers.resolvePageHead(pageHeadTitle.value, pageFrontmatter.value, siteLocaleData.value));
    const pageLang = computed(() => resolvers.resolvePageLang(pageData.value));
    // provide global computed
    app.provide(routeLocaleSymbol, routeLocale);
    app.provide(siteLocaleDataSymbol, siteLocaleData);
    app.provide(pageFrontmatterSymbol, pageFrontmatter);
    app.provide(pageHeadTitleSymbol, pageHeadTitle);
    app.provide(pageHeadSymbol, pageHead);
    app.provide(pageLangSymbol, pageLang);
    // provide global helpers
    Object.defineProperties(app.config.globalProperties, {
        $frontmatter: { get: () => pageFrontmatter.value },
        $head: { get: () => pageHead.value },
        $headTitle: { get: () => pageHeadTitle.value },
        $lang: { get: () => pageLang.value },
        $page: { get: () => pageData.value },
        $routeLocale: { get: () => routeLocale.value },
        $site: { get: () => siteData.value },
        $siteLocale: { get: () => siteLocaleData.value },
        $withBase: { get: () => withBase },
    });
    return {
        pageData,
        pageFrontmatter,
        pageHead,
        pageHeadTitle,
        pageLang,
        routeLocale,
        siteData,
        siteLocaleData,
    };
};
