// @vuepress/plugin-medium-zoom/lib/node/mediumZoomPlugin.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.mediumZoomPlugin = void 0;
const utils_1 = require("@vuepress/utils");
const mediumZoomPlugin = ({ selector = ':not(a) > img', zoomOptions = {}, delay = 500 }, app) => {
    if (app.env.isDev && app.options.bundler.endsWith('vite')) {
        // eslint-disable-next-line import/no-extraneous-dependencies
        app.options.bundlerConfig.viteOptions = require('vite').mergeConfig(app.options.bundlerConfig.viteOptions, {
            optimizeDeps: {
                exclude: ['medium-zoom'],
            },
        });
    }
    return {
        name: '@vuepress/plugin-medium-zoom',
        clientAppEnhanceFiles: utils_1.path.resolve(__dirname, '../client/clientAppEnhance.js'),
        define: {
            __MZ_SELECTOR__: selector,
            __MZ_ZOOM_OPTIONS__: zoomOptions,
            __MZ_DELAY__: delay,
        },
    };
};
exports.mediumZoomPlugin = mediumZoomPlugin;
exports.default = exports.mediumZoomPlugin;
